File: old/src/main/java/practice/Maze.java
package practice;

import java.util.LinkedList;

public class Maze {
	public boolean hasPath(int[][] maze, int[] start, int[] destination) {
		var queue = new LinkedList<int[]>();
		queue.add(start);
		var visited = new boolean[maze.length][maze[0].length];
		visited[start[0]][start[1]] = true;
		while (!queue.isEmpty()) {
			var cell = queue.poll();
			for (var direction : directions) {
				var nextX = cell[0];
				var nextY = cell[1];
				while (isValid(nextX + direction[0], nextY + direction[1], maze)
						&& maze[nextX + direction[0]][nextY + direction[1]] != 1) {
					nextX += direction[0];
					nextY += direction[1];
				}
				if (nextX == destination[0] && nextY == destination[1]) {
					return true;
				}
				if (!visited[nextX][nextY]) {
					visited[nextX][nextY] = true;
					queue.add(new int[] {nextX, nextY});
				}
			}
		}
		return false;
	}

	private boolean isValid(int x, int y, int[][] maze) {
		return (x >= 0 && x < maze.length) && (y >= 0 && y < maze[0].length);
	}

	int[][] directions = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

	public static void main(String[] args) {
		var maze =
				new int[][] {
					{0, 0, 1, 0, 0},
					{0, 0, 0, 0, 0},
					{0, 0, 0, 1, 0},
					{1, 1, 0, 1, 1},
					{0, 0, 0, 0, 0}
				};
		var start = new int[] {0, 4};
		var destination = new int[] {4, 4};
		var mazeSolver = new Maze();
		System.out.println(mazeSolver.hasPath(maze, start, destination)); // true
	}
}

File: ds-algo/src/main/kotlin/leetcode/graph/Maze.kt
package leetcode.graph

/* 10 Aug 2025 15:39 */

/** [The Maze](https://leetcode.ca/2017-04-03-490-The-Maze/) */
fun hasPath(maze: Array<IntArray>, start: IntArray, destination: IntArray): Boolean {
  val nextCells =
    directions
      .map { intArrayOf(it.first + start[0], it.second + start[1]) }
      .filter { isValid(it, maze) }
  return when {
    nextCells.isEmpty() -> start.contentEquals(destination)
    else -> {
      maze[start[0]][start[1]] = 1
      nextCells.any { hasPath(maze, it, destination) }
    }
  }
}

private val directions = listOf(-1 to 0, 1 to 0, 0 to -1, 0 to 1)

private fun isValid(cell: IntArray, maze: Array<IntArray>): Boolean =
  cell[0] in maze.indices && cell[1] in maze.first().indices && maze[cell[0]][cell[1]] == 0

fun main() {
  println(
    hasPath(
      arrayOf(
        intArrayOf(0, 0, 1, 0, 0),
        intArrayOf(0, 0, 0, 0, 0),
        intArrayOf(0, 0, 0, 1, 0),
        intArrayOf(1, 1, 0, 1, 1),
        intArrayOf(0, 0, 0, 0, 0),
      ),
      intArrayOf(0, 4),
      intArrayOf(4, 4),
    )
  )

  println(
    hasPath(
      arrayOf(
        intArrayOf(0, 0, 1, 0, 0),
        intArrayOf(0, 0, 0, 0, 0),
        intArrayOf(0, 0, 0, 1, 0),
        intArrayOf(1, 1, 0, 1, 1),
        intArrayOf(0, 0, 0, 0, 0),
      ),
      intArrayOf(0, 4),
      intArrayOf(3, 2),
    )
  )

  println(
    hasPath(
      arrayOf(
        intArrayOf(0, 0, 0, 0, 0),
        intArrayOf(1, 1, 0, 0, 1),
        intArrayOf(0, 0, 0, 0, 0),
        intArrayOf(0, 1, 0, 0, 1),
        intArrayOf(0, 1, 0, 0, 0),
      ),
      intArrayOf(4, 3),
      intArrayOf(0, 1),
    )
  )
}
